### ISSUE TITLE
Write a rule for exception throwing

### ISSUE THREAD

**Issue by andresuribe87** [TBD54566975/tbd-detekt-rules#7]
As shown in https://github.com/TBD54566975/web5-kt/pull/154#discussion_r1419773443, we need to have `@Throws` annotations, and `@throws` comments for public methods. This issue is to write a rule that enforces such behavior such behavior. 

As an added bonus, it would be amazing to traverse the call tree and enforce that all the exceptions thrown are captured in the `@Throws` annotation. I'm unsure if something like that is possible though, but it sure would be nice.

**Comment by jiyoontbd** [comment]
relevant comment from @KendallWeihe when thinking about building custom detekt rule for exceptions - https://github.com/TBD54566975/web5-kt/issues/144#issuecomment-1842789994

**Comment by KendallWeihe** [comment]
> relevant comment from @KendallWeihe when thinking about building custom detekt rule for exceptions - https://github.com/TBD54566975/web5-kt/issues/144#issuecomment-1842789994

TL;DR detekt rule which requires any publicly facing function to be wrapped in a try/catch which throws a standard `Web5Exception` (or whatevs we wanna name it) for any unknown exception case (meaning, not one of our custom exceptions)

The reason being, we don't bleed exceptions thrown from transitive dependencies, which makes this (from @andresuribe87  description) possible:

> it would be amazing to traverse the call tree and enforce that all the exceptions thrown are captured in the @Throws annotation

**Comment by KendallWeihe** [comment]
Ya know, I think you're right. As I was re-reading and typing that out earlier I had similar thoughts.

I think it _would_ solve the problem of documenting all possible exceptions, but that's the extend of it. The exceptions would still occur, and would still contain the same exception info, and may still be caused by a transitive dependency. So is anything actually gained aside from meeting the expectations of what's documented? I don't think so, and in a way it kind of obfuscates, or at a minimum it further confuses, what the actual underlying exception was.

So yeah, I'm convinced this is akin to wrapping something in gold and then pretending like the thing is solid gold.
